Use crypto lib with removed multiSigSign/verify

hewison-chris opened the pull request:
This is now using Crypto submodule which has been merged into v0.x.x

hewison-chris commented:
This will need to be updated with changes in Signature just changed in the crypto lib. Not ready to merge yet.

hewison-chris committed:
Use crypto lib returning Schnorr Signature

This includes the crypto submodule updates which include:
1. replacment of multiSigSign / Verify with sign / verify overloads
2. returning a Schnorr Signature struct (`R`, `s`) rather than a bitblob
    when signing. `R` is the commitment (ECC Point on the 25519 curve)
    `s` is the proof (ECC Scalar in the lower group : 1 .. `L`)
        where `L` = 2^252 + 27742317777372353535851937790883648493

Geod24 commented on source/agora/consensus/data/Enrollment.d:
```diff
@@ -26,7 +26,7 @@ module agora.consensus.data.Enrollment;
 import agora.common.Types;
 import agora.crypto.Hash;
-
+import agora.crypto.Schnorr: Signature;
```
Newline

hewison-chris replied:
Updated

Geod24 commented on source/agora/consensus/data/genesis/Test.d:
```diff
@@ -28,8 +28,10 @@ import agora.common.Types;
 import agora.consensus.data.Block;
 import agora.consensus.data.Enrollment;
 import agora.consensus.data.Transaction;
+import agora.crypto.ECC: Scalar, Point;
```
```suggestion
import agora.crypto.ECC;
```

hewison-chris replied:
Updated.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -648,7 +648,7 @@ extern(D):
         const Point R = r.toPoint();
         log.trace("createBlockSignature: Block signing commitment R for validator {} is {}", this.kp.address, R);
-        return Sig(R, multiSigSign(r, this.kp.secret, challenge));
+        return Signature(R, sign(this.kp.secret, R, r, challenge).s);
```
```suggestion
        return sign(this.kp.secret, R, r, challenge);
```

hewison-chris replied:
Updated.

Geod24 commented on source/agora/crypto/Key.d:
```diff
@@ -43,10 +43,6 @@ unittest
     // Another keypair won't verify it
     KeyPair other = KeyPair.random();
     assert(!other.address.verify(sign, "Hello World".representation));
-
-    // Signature can't be changed
-    sign[].ptr[0] = sign[].ptr[0] ? 0 : 1;
-    assert(!kp.address.verify(sign, "Hello World".representation));
```
??

hewison-chris replied:
Reverted but changed the signature by signing something else as it didn't compile as it was.

Geod24 commented on source/agora/flash/API.d:
```diff
@@ -16,6 +16,7 @@ module agora.flash.API;
 import agora.common.Amount;
 import agora.common.Types;
 import agora.crypto.ECC;
+import agora.crypto.Schnorr: Signature;
```
Doesn't seem used here ?

hewison-chris replied:
https://github.com/bpfkorea/agora/blob/1efa307d45090915d40ba52ef124affafd398a45/source/agora/flash/API.d#L261

Geod24 commented on source/agora/flash/Scripts.d:
```diff
@@ -24,6 +24,7 @@ import agora.flash.Types;
 import agora.script.Lock;
 import agora.script.Opcodes;
 import agora.script.Script;
+import agora.serialization.Serializer;
```
Why ? The diff only add `toBlob`

Geod24 commented on source/agora/node/BlockStorage.d:
```diff
@@ -21,6 +21,7 @@ import agora.common.BitField;
 import agora.common.Types;
 import agora.consensus.data.Block;
 import agora.crypto.Hash;
+import agora.crypto.Schnorr: Signature;
```
Unused ?

hewison-chris replied:
https://github.com/bpfkorea/agora/blob/1efa307d45090915d40ba52ef124affafd398a45/source/agora/node/BlockStorage.d#L116

Geod24 commented on source/agora/node/admin/QRCodeInterface.d:
```diff
@@ -19,6 +19,7 @@ import agora.common.Types;
 import agora.consensus.data.PreImageInfo;
 import agora.crypto.Hash;
 import agora.crypto.Key;
+import agora.crypto.Schnorr: Signature;
```
Same question, is this used ?

hewison-chris replied:
https://github.com/bpfkorea/agora/blob/1efa307d45090915d40ba52ef124affafd398a45/source/agora/node/admin/QRCodeInterface.d#L231

Geod24 commented on source/agora/registry/API.d:
```diff
@@ -16,6 +16,7 @@ module agora.registry.API;
 import agora.common.Types;
 import agora.crypto.Hash;
 import agora.crypto.Key;
+import agora.crypto.Schnorr: Signature;
```
Is this used ?

hewison-chris replied:
https://github.com/bpfkorea/agora/blob/1efa307d45090915d40ba52ef124affafd398a45/source/agora/registry/API.d#L44

Geod24 commented on source/agora/script/Lock.d:
```diff
@@ -34,7 +34,7 @@ module agora.script.Lock;
 import agora.crypto.ECC;
 import agora.crypto.Key;
-
+import agora.crypto.Schnorr: Signature;
```
Doesn't seem used

hewison-chris replied:
https://github.com/bpfkorea/agora/blob/1efa307d45090915d40ba52ef124affafd398a45/source/agora/script/Lock.d#L152

Geod24 commented on source/agora/script/Script.d:
```diff
@@ -13,8 +13,10 @@
 
 module agora.script.Script;
 
+import agora.common.Types: toBlob;
```
```suggestion
import agora.common.Types;
```

hewison-chris replied:
Updated.

Geod24 commented on source/agora/script/Script.d:
```diff
@@ -13,8 +13,10 @@
+import agora.common.Types: toBlob;
 import agora.crypto.ECC;
 import agora.crypto.Hash;
+import agora.crypto.Schnorr: Signature;
```
Unused

hewison-chris replied:
https://github.com/bpfkorea/agora/blob/1efa307d45090915d40ba52ef124affafd398a45/source/agora/script/Script.d#L352

Geod24 commented on source/agora/script/Engine.d:
```diff
@@ -20,7 +20,7 @@
 module agora.script.Engine;
 
-import agora.common.Types : Height;
+import agora.common.Types : Height, toSignature, toBlob;
```
```suggestion
import agora.common.Types;
```

hewison-chris replied:
Updated.

Geod24 commented on source/agora/script/Engine.d:
```diff
@@ -165,9 +165,9 @@ public class Engine
                 return "LockType.Key 32-byte public key in lock script is invalid";
 
-            if (unlock.bytes.length != Signature.sizeof)
+            if (unlock.bytes.length != Schnorr.Signature.sizeof)
```
I'd say just remove the rename of the import, it was only there to work around the name clash.

hewison-chris replied:
I thought it might be to document but more likely because it was a clash.

hewison-chris replied:
Updated.

Geod24 commented:
Also there's an accidental vibe-core revert.

Geod24 commented:
@hewison-chris : Regarding the "unused", I suppose it's because those modules import `agora.crypto.Types`, which used to define `Signature` ? In this case, disregard my comments about this.

hewison-chris commented:
> Also there's an accidental vibe-core revert.

Reverted

hewison-chris commented:
> @hewison-chris : Regarding the "unused", I suppose it's because those modules import `agora.crypto.Types`, which used to define `Signature` ? In this case, disregard my comments about this.

Didn't do any harm to check them anyway.

hewison-chris commented:
The CI tests fail and looks like it is with the signature. Will have to fix before we can merge.

Geod24 commented:
> The CI tests fail and looks like it is with the signature. Will have to fix before we can merge.

Hoping to get this merged today, ping me when ready.

hewison-chris commented:
> > The CI tests fail and looks like it is with the signature. Will have to fix before we can merge.
> 
> Hoping to get this merged today, ping me when ready.

Just building a compatible Faucet for the ci tests. If this works should be ready today.

Geod24 commented on source/agora/test/InvalidBlockSigByzantine.d:
```diff
@@ -76,15 +76,15 @@ private extern(C++) class BadBlockSigningNominator : TestNominator
                 // Sign with random in place of validator secret key
                 const wrong_scalar_v = Scalar.random();
-                return Sig(R, multiSigSign(r, wrong_scalar_v, challenge));
+                return Signature(R, sign(wrong_scalar_v, R, r, challenge).s);
```
```suggestion
                return sign(wrong_scalar_v, R, r, challenge);
```

hewison-chris replied:
Updated

Geod24 commented on source/agora/test/InvalidBlockSigByzantine.d:
```diff
@@ -76,15 +76,15 @@ private extern(C++) class BadBlockSigningNominator : TestNominator
                 const Scalar r = rc + challenge;
                 const Point R = r.toPoint();
-                return Sig(R, multiSigSign(r, this.kp.secret, challenge));
+                return Signature(R, sign(this.kp.secret, R, r, challenge).s);
```
```suggestion
                return sign(this.kp.secret, R, r, challenge);
```

hewison-chris replied:
Updated

Geod24 reviewed (approved):
Sweet

Geod24 commented:
> node-0_1  | 2021-03-18 05:42:19,866 Error [agora.consensus.validation.Block] - [source/agora/node/FullNode.d:343] Block: Invalid schnorr signature for 6 active validators. Sum of R mismatch

Geod24 commented:
This needs to be merged ASAP as it blocks any other PR from running the integration tests.

hewison-chris commented:
> This needs to be merged ASAP as it blocks any other PR from running the integration tests.

I am adding a full node to the Periodic catchup test to locate the problem.

hewison-chris committed:
Increase timeout for discovery

hewison-chris committed:
TEMP: exclude node 0 from height check

This is to unblock ci tests which for some unknown reason fail.
The full node node-0 reports invalid signature in the blocks fetched
from the other nodes so never accepts any blocks.

hewison-chris commented:
I added a full node to the PeriodicCatchup network test and it was able to verify the fetched blocks correctly.
The last commit skips the height assertion for node-0 to unblock ci till the issue is resolved.
```>>CI: All nodes reached target height 2 (excluding the fullnode)```

Geod24 commented:
~So why do we have this extra node ?~

Nvm I'm dumb
